### PR TITLE
Feature/match path expression

### PR DESCRIPTION
**Pull request by Shylock-Hg**
## What type of PR is this?
- [ ] bug
- [x] feature
- [ ] enhancement

#### Issue(s) number: 
Sub job of #3131 

## Checklist:
Tests:
- [ ] Unit test(positive and negative cases)
- [ ] Function test
- [ ] Performance test
- [ ] N/A

Affects:
- [ ] Documentation affected (Please add the label if documentation needs to be modified.)
- [ ] Incompatibility (If it breaks the compatibility, please describe it and add the label.）
- [ ] If it's needed to cherry-pick (If cherry-pick to some branches is required, please label the destination version(s).)
- [ ] Performance impacted: Consumes more CPU/Memory

## Release notes:

Please confirm whether to be reflected in release notes and how to describe:
> ex. Fixed the bug .....

### CODE REVIEW

File: tests/tck/ldbc/business_intelligence_workload/Read.feature [hunk 1]
```diff
@@ -214,14 +214,13 @@ Feature: LDBC Business Intelligence Workload - Read
       | person1Id | authorityScore |
 
   Scenario: 8. Related topics
```
Comment (CPWstatic) [inline tests/tck/ldbc/business_intelligence_workload/Read.feature:216]:
```
MATCH
  (tag:Tag {name: $tag})<-[:HAS_TAG]-(message:Message),
  (message)<-[:REPLY_OF]-(comment:Comment)-[:HAS_TAG]->(relatedTag:Tag)
WHERE NOT (comment)-[:HAS_TAG]->(tag)
RETURN
  relatedTag.name,
  count(DISTINCT comment) AS count
ORDER BY
  count DESC,
  relatedTag.name ASC
LIMIT 100
```

File: tests/tck/ldbc/business_intelligence_workload/Read.feature [hunk 2]
```diff
@@ -263,68 +262,42 @@ Feature: LDBC Business Intelligence Workload - Read
 
   @skip
   Scenario: 10. Central person for a tag
```
Comment (CPWstatic) [inline tests/tck/ldbc/business_intelligence_workload/Read.feature:264]:
```
MATCH (tag:Tag {name: $tag})
// score
OPTIONAL MATCH (tag)<-[interest:HAS_INTEREST]-(person:Person)
WITH tag, collect(person) AS interestedPersons
OPTIONAL MATCH (tag)<-[:HAS_TAG]-(message:Message)-[:HAS_CREATOR]->(person:Person)
         WHERE message.creationDate > $date
WITH tag, interestedPersons + collect(person) AS persons
UNWIND persons AS person
// poor man's disjunct union (should be changed to UNION + post-union processing in the future)
WITH DISTINCT tag, person
WITH
  tag,
  person,
  100 * length([(tag)<-[interest:HAS_INTEREST]-(person) | interest])
    + length([(tag)<-[:HAS_TAG]-(message:Message)-[:HAS_CREATOR]->(person) WHERE message.creationDate > $date | message])
  AS score
OPTIONAL MATCH (person)-[:KNOWS]-(friend)
WITH
  person,
  score,
  100 * length([(tag)<-[interest:HAS_INTEREST]-(friend) | interest])
    + length([(tag)<-[:HAS_TAG]-(message:Message)-[:HAS_CREATOR]->(friend) WHERE message.creationDate > $date | message])
  AS friendScore
RETURN
  person.id,
  score,
  sum(friendScore) AS friendsScore
ORDER BY
  score + friendsScore DESC,
  person.id ASC
LIMIT 100
```

File: tests/tck/ldbc/business_intelligence_workload/Read.feature [hunk 3]
```diff
@@ -531,31 +504,31 @@ Feature: LDBC Business Intelligence Workload - Read
       | messageCount | personCount |
 
   Scenario: 19. Stranger’s interaction
```
Comment (CPWstatic) [inline tests/tck/ldbc/business_intelligence_workload/Read.feature:506]:
```
MATCH
  (:TagClass {name: $tagClass1})<-[:HAS_TYPE]-(:Tag)<-[:HAS_TAG]-
  (forum1:Forum)-[:HAS_MEMBER]->(stranger:Person)
WITH DISTINCT stranger
MATCH
  (:TagClass {name: $tagClass2})<-[:HAS_TYPE]-(:Tag)<-[:HAS_TAG]-
  (forum2:Forum)-[:HAS_MEMBER]->(stranger)
WITH DISTINCT stranger
MATCH
  (person:Person)<-[:HAS_CREATOR]-(comment:Comment)-[:REPLY_OF*]->(message:Message)-[:HAS_CREATOR]->(stranger)
WHERE person.birthday > $date
  AND person <> stranger
  AND NOT (person)-[:KNOWS]-(stranger)
  AND NOT (message)-[:REPLY_OF*]->(:Message)-[:HAS_CREATOR]->(stranger)
RETURN
  person.id,
  count(DISTINCT stranger) AS strangersCount,
  count(comment) AS interactionCount
ORDER BY
  interactionCount DESC,
  person.id ASC
LIMIT 100
```

File: tests/tck/ldbc/interactive_workload/ComplexReads.feature [hunk 4]
```diff
@@ -82,43 +82,43 @@ Feature: LDBC Interactive Workload - Complex Reads
       | personId | personFirstName | personLastName | messageId | messageContent | messageCreationDate |
 
   Scenario: 3. Friends and friends of friends that have been to given countries
```
Comment (CPWstatic) [inline tests/tck/ldbc/interactive_workload/ComplexReads.feature:84]:
```
MATCH (countryX:Country {name: $countryXName }),
      (countryY:Country {name: $countryYName }),
      (person:Person {id: $personId })
WITH person, countryX, countryY
LIMIT 1
MATCH (city:City)-[:IS_PART_OF]->(country:Country)
WHERE country IN [countryX, countryY]
WITH person, countryX, countryY, collect(city) AS cities
MATCH (person)-[:KNOWS*1..2]-(friend)-[:IS_LOCATED_IN]->(city)
WHERE NOT person=friend AND NOT city IN cities
WITH DISTINCT friend, countryX, countryY
MATCH (friend)<-[:HAS_CREATOR]-(message),
      (message)-[:IS_LOCATED_IN]->(country)
WHERE $endDate > message.creationDate >= $startDate AND
      country IN [countryX, countryY]
WITH friend,
     CASE WHEN country=countryX THEN 1 ELSE 0 END AS messageX,
     CASE WHEN country=countryY THEN 1 ELSE 0 END AS messageY
WITH friend, sum(messageX) AS xCount, sum(messageY) AS yCount
WHERE xCount>0 AND yCount>0
RETURN friend.id AS friendId,
       friend.firstName AS friendFirstName,
       friend.lastName AS friendLastName,
       xCount,
       yCount,
       xCount + yCount AS xyCount
ORDER BY xyCount DESC, friendId ASC
LIMIT 20
```

File: tests/tck/ldbc/interactive_workload/ComplexReads.feature [hunk 5]
```diff
@@ -254,34 +253,34 @@ Feature: LDBC Interactive Workload - Complex Reads
     Then the result should be, in any order:
       | personId | personFirstName | personLastName | messageId | messageContent | messageCreationDate |
 
+  @skip
   Scenario: 10. Friend recommendation
```
Comment (CPWstatic) [inline tests/tck/ldbc/interactive_workload/ComplexReads.feature:257]:
```
MATCH (person:Person {id: $personId})-[:KNOWS*2..2]-(friend),
       (friend)-[:IS_LOCATED_IN]->(city:City)
WHERE NOT friend=person AND
      NOT (friend)-[:KNOWS]-(person)
WITH person, city, friend, datetime({epochMillis: friend.birthday}) as birthday
WHERE  (birthday.month=$month AND birthday.day>=21) OR
        (birthday.month=($month%12)+1 AND birthday.day<22)
WITH DISTINCT friend, city, person
OPTIONAL MATCH (friend)<-[:HAS_CREATOR]-(post:Post)
WITH friend, city, collect(post) AS posts, person
WITH friend,
     city,
     size(posts) AS postCount,
     size([p IN posts WHERE (p)-[:HAS_TAG]->()<-[:HAS_INTEREST]-(person)]) AS commonPostCount
RETURN friend.id AS personId,
       friend.firstName AS personFirstName,
       friend.lastName AS personLastName,
       commonPostCount - (postCount - commonPostCount) AS commonInterestScore,
       friend.gender AS personGender,
       city.name AS personCityName
ORDER BY commonInterestScore DESC, personId ASC
LIMIT 10
```

Comment (Shylock-Hg) [reply tests/tck/ldbc/interactive_workload/ComplexReads.feature:257]: Variable defined in list comprehension is not supported.

**Comment by Shylock-Hg** [comment]
The origin LALR(1) parser benchmark:
```
/*
Architecture:        x86_64
CPU op-mode(s):      32-bit, 64-bit
Byte Order:          Little Endian
Address sizes:       46 bits physical, 57 bits virtual
CPU(s):              128
On-line CPU(s) list: 0-127
Thread(s) per core:  2
Core(s) per socket:  32
Socket(s):           2
NUMA node(s):        2
Vendor ID:           GenuineIntel
CPU family:          6
Model:               106
Model name:          Intel(R) Xeon(R) Platinum 8352Y CPU @ 2.20GHz
============================================================================
/home/shylock.huang/nebula-tmp/src/parser/test/ParserBenchmark.cpprelative  time/iter  iters/s
============================================================================
SimpleQuery(1_thread)                                        1.70us  587.21K
SimpleQuery(2_thread)                             96.72%     1.76us  567.92K
SimpleQuery(4_thread)                             97.35%     1.75us  571.67K
SimpleQuery(8_thread)                             96.84%     1.76us  568.66K
SimpleQuery(16_thread)                            95.77%     1.78us  562.36K
SimpleQuery(32_thread)                            53.77%     3.17us  315.74K
SimpleQuery(48_thread)                            49.68%     3.43us  291.73K
----------------------------------------------------------------------------
ComplexQuery(1_thread)                                       9.94us  100.57K
ComplexQuery(2_thread)                            98.73%    10.07us   99.30K
ComplexQuery(4_thread)                           101.62%     9.78us  102.21K
ComplexQuery(8_thread)                            90.47%    10.99us   90.99K
ComplexQuery(16_thread)                           97.32%    10.22us   97.87K
ComplexQuery(32_thread)                           69.31%    14.34us   69.71K
ComplexQuery(48_thread)                           61.30%    16.22us   61.66K
----------------------------------------------------------------------------
MatchConflictQuery(1_thread)                                 3.88us  257.41K
MatchConflictQuery(2_thread)                      99.74%     3.90us  256.73K
MatchConflictQuery(4_thread)                      96.16%     4.04us  247.52K
MatchConflictQuery(8_thread)                      79.93%     4.86us  205.74K
MatchConflictQuery(16_thread)                     79.57%     4.88us  204.81K
MatchConflictQuery(32_thread)                     56.12%     6.92us  144.47K
MatchConflictQuery(48_thread)                     50.41%     7.71us  129.75K
============================================================================
*/
```

**Review by CPWstatic — APPROVED** [review]

**Review by yixinglu — APPROVED** [review]

**Pull request merged**